modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointsSelfTest.java

```diff
@@ -0,0 +1,290 @@
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.transactions;
+
+import java.util.concurrent.Callable;
+import org.apache.ignite.IgniteCache;
+import org.apache.ignite.configuration.CacheConfiguration;
+import org.apache.ignite.internal.processors.cache.GridCacheAbstractSelfTest;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.transactions.Transaction;
+import org.apache.ignite.transactions.TransactionConcurrency;
+import org.apache.ignite.transactions.TransactionIsolation;
+
+/**
+ *
+ */
+public class TxSavepointsSelfTest extends GridCacheAbstractSelfTest {
+    /** {@inheritDoc} */
+    @Override protected int gridCount() {
+        return 3;
+    }
+
+    /**
+     * Tests savepoint.
+     */
+    public void testSavepoints() {
+        for (TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (TransactionIsolation isolation : TransactionIsolation.values()) {
+                jcache().put("key", 0);
+
+                try (Transaction tx = grid(0).transactions().txStart(concurrency, isolation)) {
+                    jcache().put("key", 1);
+
+                    tx.savepoint("s1");
+
+                    jcache().put("key", 2);
+
+                    tx.savepoint("s2");
+
+                    jcache().put("key", 3);
+
+                    tx.savepoint("s3");
+
+                    tx.rollbackToSavepoint("s2");
+
+                    assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                        (Integer) 2, jcache().get("key"));
+
+                    tx.commit();
+                }
+
+                assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                    (Integer) 2, jcache().get("key"));
+            }
+    }
+
+    /**
+     * Tests valid and invalid rollbacks to savepoint.
+     */
+    public void testFailRollbackToSavepoint() {
+        for (final TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (final TransactionIsolation isolation : TransactionIsolation.values()) {
+                GridTestUtils.assertThrows(
+                    log, new Callable<Boolean>() {
+                        @Override public Boolean call() throws Exception {
+                            jcache().put("key", 0);
+
+                            try (Transaction tx = grid(0).transactions().txStart(concurrency, isolation)) {
+                                jcache().put("key", 1);
+
+                                tx.savepoint("s1");
+
+                                jcache().put("key", 2);
+
+                                tx.savepoint("s2");
+
+                                jcache().put("key", 3);
+
+                                tx.savepoint("s3");
+
+                                tx.rollbackToSavepoint("s2");
+
+                                assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
```

Comment:
may be move all such asserts to one method and reuse it ? And may be change message to smth more informative

Reply:
I'm not sure that replacing it by method with 4 arguments is good idea.